pinot-controller/src/main/java/org/apache/pinot/controller/util/TableTierReader.java

```diff
@@ -0,0 +1,123 @@
+  private final Executor _executor;
+  private final HttpConnectionManager _connectionManager;
+  private final PinotHelixResourceManager _helixResourceManager;
+
+  public TableTierReader(Executor executor, HttpConnectionManager connectionManager,
+      PinotHelixResourceManager helixResourceManager) {
+    _executor = executor;
+    _connectionManager = connectionManager;
+    _helixResourceManager = helixResourceManager;
+  }
+
+  /**
+   * Get the segment storage tiers for the given table. The servers or segments not responding the request are
+   * recorded in the result to be checked by caller.
+   *
+   * @param tableNameWithType table name with type
+   * @param timeoutMs timeout for reading segment tiers from servers
+   * @return details of segment storage tiers for the given table
+   */
+  public TableTierDetails getTableTierDetails(String tableNameWithType, @Nullable String segmentName, int timeoutMs)
+      throws InvalidConfigException {
+    Map<String, List<String>> serverToSegmentsMap = new HashMap<>();
+    if (segmentName == null) {
+      serverToSegmentsMap.putAll(_helixResourceManager.getServerToSegmentsMap(tableNameWithType));
+    } else {
+      List<String> segmentInList = Collections.singletonList(segmentName);
+      for (String server : _helixResourceManager.getServers(tableNameWithType, segmentName)) {
+        serverToSegmentsMap.put(server, segmentInList);
+      }
+    }
+    BiMap<String, String> endpoints = _helixResourceManager.getDataInstanceAdminEndpoints(serverToSegmentsMap.keySet());
+    ServerTableTierReader serverTableTierReader = new ServerTableTierReader(_executor, _connectionManager);
+    Map<String, TableTierInfo> serverToTableTierInfoMap =
+        serverTableTierReader.getTableTierInfoFromServers(endpoints, tableNameWithType, timeoutMs);
+
+    TableTierDetails tableTierDetails = new TableTierDetails(tableNameWithType);
+    for (Map.Entry<String, List<String>> entry : serverToSegmentsMap.entrySet()) {
+      String server = entry.getKey();
+      TableTierInfo tableTierInfo = serverToTableTierInfoMap.get(server);
+      if (tableTierInfo == null) {
+        tableTierDetails._missingServers.add(server);
+        continue;
+      }
+      Map<String, String> segmentTiers = tableTierInfo.getSegmentTiers();
+      for (String expectedSegment : entry.getValue()) {
+        if (!segmentTiers.containsKey(expectedSegment)) {
+          tableTierDetails._missingSegments.computeIfAbsent(server, (k) -> new HashSet<>()).add(expectedSegment);
+        } else {
+          tableTierDetails._segmentTiers.computeIfAbsent(expectedSegment, (k) -> new HashMap<>())
+              .put(server, segmentTiers.get(expectedSegment));
+        }
+      }
+    }
+    return tableTierDetails;
+  }
+
+  // This class aggregates the TableTierInfo returned from multi servers.
+  public static class TableTierDetails {
+    private final String _tableName;
+    private final Set<String> _missingServers = new HashSet<>();
```

Comment:
splitting these details (missing servers, missing segments, responses from happy path) is nice when you want to look at just one of these. But I'm wondering, if this will make it too many places to look at to get the complete picture.
wdyt about having all the info in _segmentTiers? the missingSegments and missingServers can be there in addition. so what i mean is, for the missing segment, put it in `segmentTiers` map, as segment->server,null/empty ? And also have an entry for any segment that was supposed to be found on the missingServer.
If there's some gotcha in doing the above (like local tier is also represented as null?), then maybe just make `missingSegments` map as segment->server, so it is easier to read the 2 together?

Reply:
lemme know your thoughts?
```
{
  "segmentTiers": {
    "seg01": {
      "server1": null,
      "server0": null
    },
    "seg02": {
      "server0": "someTier"
      "server2": "NO_RESPONSE_FROM_SERVER"
    },
    "seg03": {
      "server4": "SEGMENT_MISSED_ON_SERVER"
    }
  },
  "tableName": "myTable_OFFLINE"
}
```